ReactNativeClient/lib/services/searchengine/filterParser.ts

```diff
@@ -0,0 +1,101 @@
+interface Filter {
+	relation: string,
+	name: string,
+	value: string
+}
+
+interface Term {
+  relation: string;
+  value: string;
+}
+
+const parseQuery = (query: string): Array<Filter> => {
+	const terms = [];
+
+	// tag:123 tag:234 or some query -> [["tag", "123"], ["tag", "234"], ["_", "or"], ["_", "some"], ["_", "query"]]
+	let inQuote = false;
+	let currentCol = '_';
+	let currentTerm = '';
+	for (let i = 0; i < query.length; i++) {
+		const c = query[i];
+
+		if (c === '"') {
+			currentTerm += c; // keep the quotes
+			if (inQuote) {
+				terms.push([currentCol, currentTerm]);
+				currentTerm = '';
+				inQuote = false;
+			} else {
+				inQuote = true;
+			}
+			continue;
+		}
+
+		if (c === ' ' && !inQuote) {
+			if (!currentTerm) continue;
+			terms.push([currentCol, currentTerm]);
+			currentCol = '_';
+			currentTerm = '';
+			continue;
+		}
+
+		if (c === ':' && !inQuote) {
+			currentCol = currentTerm;
+			currentTerm = '';
+			continue;
+		}
+
+		currentTerm += c;
+	}
+	if (currentTerm) terms.push([currentCol, currentTerm]);
+
+
+	let relation = 'AND'; // default relation
+	const result: Array<Filter> = [];
+	for (let i = 0; i < terms.length; i++) {
+		if (terms[i][0] !== '_') {
+			// Hopefully a valid filter
+			// TO DO : handle negation
+			result.push({ relation: relation, name: terms[i][0], value: terms[i][1] });
+			relation = 'AND'; // reset to default
+		} else {
+			// could be AND or OR or text to fts search
+			if (terms[i][1].toUpperCase() !== 'AND' && terms[i][1].toUpperCase() !== 'OR') {
+				// this is text
+				result.push({ relation: relation, name: 'text', value: terms[i][1] });
+				relation = 'AND'; // reset to default
+			} else {
+				// this is a relation for the next term;
+				relation = terms[i][1].toUpperCase(); // OR
+			}
+		}
```

Comment:
I think the parsed query format should be changed to an AST to better show the relation between terms, and for more flexibility. At the moment the relation is associated with one term, but it should be between two or more terms. For example for `title:abc title:efg` it could be something like this:

```
{
	relation: "AND",
	terms: [
		{
			"title": "abcd",
		},
		{
			"title": "efg",
		},
	],
}
```

Doesn't have to be exactly like this but you see the idea. Perhaps have a look at existing AST and how various structures (such as SQL) are represented.

Reply:
https://discourse.joplinapp.org/t/do-we-need-an-ast/8985